controllers/consumer_controller.go

```diff
@@ -265,54 +175,173 @@ func (r *ConsumerReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 			"deployment %s: updated", deploy.Name,
 		)
 	}
-	for _, origDeploy := range toEstimateInstances {
-		deploy := origDeploy.DeepCopy()
-		partition := helpers.ParsePartitionAnnotation(deploy.Annotations[partitionAnnotation])
-		if partition == nil {
-			log.Error(nil, "failed to parse annotation with partition number.")
+
+	for _, origDeploy := range co.toUpdateInstances {
+		updateDeploy(origDeploy)
+	}
+	for _, origDeploy := range co.toEstimateInstances {
+		updateDeploy(origDeploy)
+	}
+	return result, nil
+}
+
+type consumerOperator struct {
+	consumer *konsumeratorv1alpha1.Consumer
+	mp       providers.MetricsProvider
+	log      logr.Logger
+
+	missingIds          []int32
+	pausedIds           []int32
+	runningIds          []int32
+	laggingIds          []int32
+	outdatedIds         []int32
+	toRemoveInstances   []*appsv1.Deployment
+	toUpdateInstances   []*appsv1.Deployment
+	toEstimateInstances []*appsv1.Deployment
+}
+
+func newConsumerOperator(log logr.Logger, consumer *konsumeratorv1alpha1.Consumer, managedDeploys v1.DeploymentList) (*consumerOperator, error) {
+	hash, err := hashstructure.Hash(consumer.Spec.DeploymentTemplate, nil)
+	if err != nil {
+		return nil, err
+	}
+	consumer.Status.ObservedGeneration = helpers.Ptr2Int64(int64(hash))
+	co := &consumerOperator{
+		consumer: consumer,
+		log:      log,
+	}
+	co.mp = co.newMetricsProvider()
```

Comment:
why not set this in the initialization, one line above?
```
co := &consumerOperator{
		consumer: consumer,
		log:      log,
		mp:  co.newMetricsProvider(),
	}

```

Reply:
it has dependencies like `log` and `consumer`
I agree it looks a bit confusing though